_posts/2021-01-13-hapi-fhir.md

```diff
@@ -0,0 +1,454 @@
+The interceptor works by allowing you to declare permission based on an individual request coming in.
+
+  
+
+The AuthorizationInterceptor is used by subclassing it and then registering your subclass with the RestfulServer. 
+The following example shows a subclassed interceptor implementing some basic rules:
+
+```java
+@SuppressWarnings("ConstantConditions")
+public class PatientAndAdminAuthorizationInterceptor extends AuthorizationInterceptor {
+
+    @Override
+    public  List<IAuthRule> buildRuleList(RequestDetails  theRequestDetails) {
+        // Process authorization header - The following is a fake
+        // implementation. Obviously we'd want something more real
+        // for a production scenario.
+        //
+        // In this basic example we have two hardcoded bearer tokens,
+        // one which is for a user that has access to one patient, and
+        // another that has full access.
+        IdType  userIdPatientId = null;
+        boolean  userIsAdmin = false;
+        String  authHeader = theRequestDetails.getHeader("Authorization");
+        if ("Bearer dfw98h38r".equals(authHeader)) {
+            // This user has access only to Patient/1 resources
+            userIdPatientId = new  IdType("Patient", 1L);
+        } else  if ("Bearer 39ff939jgg".equals(authHeader)) {
+            // This user has access to everything
+            userIsAdmin = true;
+        } else {
+            // Throw an HTTP 401
+            throw  new  AuthenticationException("Missing or invalid Authorization header value");
+        }
+        // If the user is a specific patient, we create the following rule chain:
+        // Allow the user to read anything in their own patient compartment
+        // Allow the user to write anything in their own patient compartment
+        // If a client request doesn't pass either of the above, deny it
+        if (userIdPatientId != null) {
+            return new RuleBuilder()
+                    .allow().read().allResources().inCompartment("Patient", userIdPatientId).andThen()
+                    .allow().write().allResources().inCompartment("Patient", userIdPatientId).andThen()
+                    .denyAll()
+                    .build();
+        }
+        // If the user is an admin, allow everything
+        if (userIsAdmin) {
+            return new RuleBuilder()
+                    .allowAll()
+                    .build();
+        }
+        // By default, deny everything. This should never get hit, but it's 
+        // good to be defensive
+        return new RuleBuilder()
+                .denyAll()
+                ´´.build();
+    }
+}
+```
+
+  
+
+## Conclusion
+The FHIR specification has been providing enhancements at various levels for several years, 
+from application development to inter-application integration.
+
+We can conclude that setting up a FHIR server is made easy by the HAPI libraries. 
```

Comment:
"... setting up a FHIR server in Java is made easy  by using the HAPI libraries."